File: packages/revocation/ts/src/inmem.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// In-memory RevocationStore with per-entry TTLs and bounded capacity.

import type { RevocationStore } from './iface.js'

interface Entry {
  expiresAtMonoMs: number
}

export class InMemoryRevocationStore implements RevocationStore {
  private readonly entries = new Map<string, Entry>()
  private readonly delegationEpochs = new Map<string, Entry & { epoch: number }>()
  private readonly defaultTtlMs: number
  private readonly maxEntries: number

  constructor(opts: { defaultTtlMs?: number; maxEntries?: number } = {}) {
    this.defaultTtlMs = opts.defaultTtlMs ?? 24 * 60 * 60 * 1000
    this.maxEntries = opts.maxEntries ?? 100_000
  }

  isRevoked(sid: string): boolean {
    const entry = this.entries.get(sid)
    if (!entry) return false
    if (entry.expiresAtMonoMs <= performance.now()) {
      this.entries.delete(sid)
      return false
    }
    return true
  }

  markRevoked(sid: string, ttlMs?: number): void {
    if (this.entries.size >= this.maxEntries) {
      this.reapExpired()
      if (this.entries.size >= this.maxEntries) throw new Error('Revocation store capacity exceeded')
    }
    this.entries.set(sid, { expiresAtMonoMs: performance.now() + (ttlMs ?? this.defaultTtlMs) })
  }

  currentDelegationEpoch(zoneId: string): number {
    const entry = this.delegationEpochs.get(zoneId)
    if (!entry) return 0
    if (entry.expiresAtMonoMs <= performance.now()) {
      this.delegationEpochs.delete(zoneId)
      return 0
    }
    return entry.epoch
  }

  markDelegationEpoch(zoneId: string, epoch: number, ttlMs?: number): void {
    const current = this.currentDelegationEpoch(zoneId)
    if (epoch <= current) return
    this.delegationEpochs.set(zoneId, {
      epoch,
      expiresAtMonoMs: performance.now() + (ttlMs ?? this.defaultTtlMs),
    })
  }

  size(): number {
    return this.entries.size
  }

  private reapExpired(): void {
    const now = performance.now()
    for (const [sid, entry] of this.entries) {
      if (entry.expiresAtMonoMs <= now) this.entries.delete(sid)
    }
  }

}

File: packages/revocation/go/inmem.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// In-memory Store with per-entry TTLs.

package revocation

import (
	"sync"
	"time"
)

// InMemoryStore is a process-local Store backed by a TTL map.
type InMemoryStore struct {
	mu      sync.Mutex
	entries map[string]time.Time
	epochs  map[string]epochEntry
	defTTL  time.Duration
}

type epochEntry struct {
	expiresAt time.Time
	epoch     int64
}

// NewInMemoryStore returns a Store using defaultTTL for entries that omit a TTL.
func NewInMemoryStore(defaultTTL time.Duration) *InMemoryStore {
	if defaultTTL <= 0 {
		defaultTTL = 24 * time.Hour
	}
	return &InMemoryStore{entries: map[string]time.Time{}, epochs: map[string]epochEntry{}, defTTL: defaultTTL}
}

// IsRevoked reports whether anchorID is currently revoked, evicting expired entries.
func (s *InMemoryStore) IsRevoked(anchorID string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	expiresAt, ok := s.entries[anchorID]
	if !ok {
		return false
	}
	if !time.Now().Before(expiresAt) {
		delete(s.entries, anchorID)
		return false
	}
	return true
}

// MarkRevoked records anchorID as revoked for ttl, falling back to the default TTL when zero.
func (s *InMemoryStore) MarkRevoked(anchorID string, ttl time.Duration) error {
	if ttl <= 0 {
		ttl = s.defTTL
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.entries[anchorID] = time.Now().Add(ttl)
	return nil
}

// CurrentDelegationEpoch returns the recorded delegation graph epoch for zoneID, evicting expired entries.
func (s *InMemoryStore) CurrentDelegationEpoch(zoneID string) int64 {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.currentEpochLocked(zoneID)
}

// MarkDelegationEpoch records epoch for zoneID when it advances past the current value.
func (s *InMemoryStore) MarkDelegationEpoch(zoneID string, epoch int64, ttl time.Duration) error {
	if ttl <= 0 {
		ttl = s.defTTL
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if epoch <= s.currentEpochLocked(zoneID) {
		return nil
	}
	s.epochs[zoneID] = epochEntry{expiresAt: time.Now().Add(ttl), epoch: epoch}
	return nil
}

func (s *InMemoryStore) currentEpochLocked(zoneID string) int64 {
	entry, ok := s.epochs[zoneID]
	if !ok {
		return 0
	}
	if !time.Now().Before(entry.expiresAt) {
		delete(s.epochs, zoneID)
		return 0
	}
	return entry.epoch
}
